proposals/0000-decl-invis-binders.rst

```diff
@@ -0,0 +1,440 @@
+
+``T1`` is currently legal, yielding ``T :: forall a. Maybe a``. The general
+rule is that the free variables of a *top-level* kind annotation on the RHS are
+brought into scope implicitly, and will be quantified in the final kind of the
+type constructor.
+
+In constrast, ``T2`` is currently illegal, because the kind annotation is not
+at the top level.
+
+We propose to drop this exotic form of implicit quantification from the
+language. Both ``T1`` and ``T2`` would become illegal, but with ``@k``-binders
+the programmer can rewrite them as follows::
+
+  type T1 @a = 'Nothing :: Maybe a
+  type T2 @a = 'Just ('Nothing :: Maybe a)
+
+This way all the variables occurring on the RHS are bound on the LHS.
+We exploit the new syntax to allow a nice, simple, uniform scoping rule.
+To cite the User's Guide, "The reason for this exception [the strange, ad-hoc
+rule about top-level kind annotations] is that there may be no other way to
+bind k".
+
+The same rule applies to type family instances::
+
+   type family F a :: k
+
+   type instance F Int = Any :: j -> j
+
+(where ``Any :: forall p. p``). This type instance is currently legal, but
+under this proposal ``j`` would not be in scope. You would have to write::
+
+   type instance F @(j->j) Int = Any :: j -> j
+
+In type instances, this form is already allowed today.
+
+Argument 5: Arity Inference
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+*Arity* is a property of type synonyms and type families that determines how
+many arguments are required at use sites (partial application is not allowed).
+The notion of arity is described in more detail in `section 6.4.9.2.1. "Type family
+declarations" <https://downloads.haskell.org/~ghc/9.2.1-alpha2/docs/html/users_guide/exts/type_families.html#type-family-declarations>`_
+of the User's Guide.
+
+Importantly, arity cannot be determined by looking at the kind of a type
+constructor. Consider ``F`` declared as follows::
+
+  type F :: Type -> forall k. Maybe k
+  type family F x
+
+The compiler can either assign it the arity of 1 or 2, and this choice will
+determine whether the equations of the type family can pattern match on ``k``.
+This will also determine whether a higher-kinded usage of ``F`` is possible.
+
+Thus, both arities can make sense depending on intended usage. Currently, GHC
+expects the programmer to employ an unsightly technique to specify the arity.
+By default, arity inference tries to include as many forall-bound variables as
+possible, to maximize the expressivity at definition site (at the cost of
+higher-kinded usage). However, the user may opt out by duplicating the return
+kind of the type family in its header::
+
+  -- arity = 2
+  type F :: Type -> forall k. Maybe k
+  type family F x
+
+  -- arity = 1
+  type G :: Type -> forall k. Maybe k
+  type family G x :: forall k. Maybe k
+
+With ``@``-binders we can do the opposite. We propose that by default, arity
+inference would include as few forall-bound variables as possible, to allow
```

Comment:
More than that, the arity would no longer be strangely influenced by the return kind signature.  Instead, you can simply count the binders to the left the `::` (if any).

Sigh: not quite.  In
```
F :: forall k. k -> Type
type family F a
```
F has arity 2, not 1.